scripts/pm/hf_from_gcs.py

```diff
@@ -0,0 +1,224 @@
+
+
+def upload_to_huggingface(local_path, repo_id, revision):
+    """Upload a local directory to Hugging Face as a specific revision."""
+    try:
+        print(f"Uploading checkpoint to Hugging Face as revision: {revision}")
+        # Make sure huggingface_hub is installed
+        subprocess.run(["pip", "install", "huggingface_hub"], check=True, capture_output=True)
+
+        # Use the Hugging Face CLI to upload
+        cmd = [
+            "python",
+            "-m",
+            "huggingface_hub",
+            "upload",
+            local_path,
+            "--repo-id",
+            repo_id,
+            "--revision",
+            f"step-{revision}",
+            "--commit-message",
+            f"Upload checkpoint for step {revision}",
+        ]
+        result = subprocess.run(cmd, check=True, capture_output=True, text=True)
+        print(f"Upload completed successfully.")
+        print(result.stdout)
+        return True
+    except subprocess.CalledProcessError as e:
+        print(f"Error uploading to Hugging Face: {e}")
+        print(f"stdout: {e.stdout}")
+        print(f"stderr: {e.stderr}")
+        return False
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Upload checkpoints from GCS to Hugging Face")
+    parser.add_argument(
+        "--repo-id", required=True, help='Target Hugging Face repository ID (e.g., "username/model-name")'
+    )
+    parser.add_argument("--dry-run", action="store_true", help="Only list checkpoints without uploading")
+    args = parser.parse_args()
+
+    # Initialize the GCS client
+    client = storage.Client()
+
+    # Collect all step directories
+    all_step_dirs = []
+
+    # Main execution
+    for directory in dirs:
+        try:
+            step_dirs = list_gcs_directories(directory)
+            all_step_dirs.extend(step_dirs)
+        except Exception as e:
+            print(f"Error listing {directory}: {e}")
+
+    # Sort all step directories by step number
+    if all_step_dirs:
+        all_step_dirs.sort(key=lambda x: x[1])
+
+        # Print sorted step directories
+        print("\nAll step directories sorted by step number:")
+        print("-" * 50)
+        for full_path, step_number in all_step_dirs:
+            print(f"- {full_path}")
+
+        print(f"\nTotal: {len(all_step_dirs)} step directories")
+
+        # Upload to Hugging Face
+        if not args.dry_run:
+            print(f"\nUploading to Hugging Face repo: {args.repo_id}")
+
+            for full_path, step_number in all_step_dirs:
+                # Create a temporary directory for downloading
+                with tempfile.TemporaryDirectory() as temp_dir:
+                    print(f"\nProcessing step {step_number} from {full_path}")
+
+                    # Download from GCS
+                    if download_from_gcs(full_path, temp_dir):
+                        # Upload to HF
+                        if upload_to_huggingface(temp_dir, args.repo_id, step_number):
```

Comment:
can we make this a bit more idempotent? Ideally we wouldn't need to download from GCS again if we finished one upload

Reply:
Yes, I'll check the repo to make sure such a commit doesn't already exist so if you run the script multiple times, it'll skip already uploaded commits!